telemetry/prometheus/src/main/scala/io/buoyant/telemetry/prometheus/PrometheusTelemeter.scala

```diff
@@ -0,0 +1,113 @@
+package io.buoyant.telemetry.prometheus
+
+import com.twitter.finagle.Service
+import com.twitter.finagle.http.{MediaType, Request}
+import com.twitter.finagle.stats.NullStatsReceiver
+import com.twitter.finagle.tracing.NullTracer
+import com.twitter.util.{Awaitable, Closable, Future}
+import io.buoyant.admin.Admin
+import io.buoyant.telemetry.{Metric, MetricsTree, Telemeter}
+
+/**
+ * This telemeter exposes metrics data in the Prometheus format, served on
+ * an admin endpoint.  It does not provide a StatsReceiver or Tracer.  It reads
+ * histogram summaries directly off of the MetricsTree and assumes that stats
+ * are being snapshotted at some appropriate interval.
+ */
+class PrometheusTelemeter(metrics: MetricsTree) extends Telemeter with Admin.WithHandlers {
+
+  private[prometheus] val handler = Service.mk { request: Request =>
+    val response = request.response
+    response.mediaType = MediaType.Txt
+    val sb = new StringBuilder()
+    writeMetrics(metrics, sb)
+    response.contentString = sb.toString
+    Future.value(response)
+  }
+
+  val adminHandlers: Seq[Admin.Handler] = Seq(
+    Admin.Handler("/admin/metrics/prometheus", handler)
+  )
+
+  val stats = NullStatsReceiver
+  def tracer = NullTracer
+  def run(): Closable with Awaitable[Unit] = Telemeter.nopRun
+
+  private[this] val disallowedChars = "[^a-zA-Z0-9:]".r
+  private[this] def escapeKey(key: String) = disallowedChars.replaceAllIn(key, "_")
+
+  private[this] def formatLabels(labels: Seq[(String, String)]): String =
+    if (labels.nonEmpty) {
+      labels.map {
+        case (k, v) =>
+          s"""$k="$v""""
+      }.mkString("{", ", ", "}")
+    } else {
+      ""
+    }
+
+  private[this] def labelExists(labels: Seq[(String, String)], name: String) =
+    labels.exists(_._1 == name)
+
+  private[this] def writeMetrics(
+    tree: MetricsTree,
+    sb: StringBuilder,
+    prefix0: Seq[String] = Nil,
+    labels0: Seq[(String, String)] = Nil
+  ): Unit = {
+
+    // Re-write elements out of the prefix into labels
+    val (prefix1, labels1) = prefix0 match {
+      case Seq("rt", router) if !labelExists(labels0, "rt") =>
+        (Seq("rt"), labels0 :+ ("rt" -> router))
+      case Seq("rt", "dst", "path", path) if !labelExists(labels0, "dst_path") =>
+        (Seq("rt", "dst_path"), labels0 :+ ("dst_path" -> path))
+      case Seq("rt", "dst", "id", id) if !labelExists(labels0, "dst_id") =>
+        (Seq("rt", "dst_id"), labels0 :+ ("dst_id" -> id))
+      case Seq("rt", "dst_id", "path", path) if !labelExists(labels0, "dst_path") =>
+        (Seq("rt", "dst_id"), labels0 :+ ("dst_path" -> path))
```

Comment:
I'd like to add an additional scope to the metric name in this case, where the path is scoped by id.  With the current behavior, you're producing two stats that have the same metric name, with just one additional label to differentiate. For instance:

```
rt:dst_id:requests{rt="external", dst_id="$/inet/web/7000", dst_path="svc/192.168.99.100:8080", type="counter"} 455
rt:dst_id:requests{rt="external", dst_id="$/inet/web/7000", type="counter"} 455
```

I'm worried that with this setup it would be too easy to double count the stats from "$/inet/web/7000". You would need to setup a query that explicitly drops to deduplicates dst_path.  Instead, I would prefer for these two stats to end up as:

```
rt:dst_id:dst_path:requests{rt="external", dst_id="$/inet/web/7000", dst_path="svc/192.168.99.100:8080", type="counter"} 455
rt:dst_id:requests{rt="external", dst_id="$/inet/web/7000", type="counter"} 455
```

And in order to do that, I think you can just change line 68 to:

```
        (Seq("rt", "dst_id", "dst_path"), labels0 :+ ("dst_path" -> path))
```

Put another way, I think it would be a good rule of thumb to expect that every custom label produced by this code also shows up as a segment in the metric name.